ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardChain.java

```diff
@@ -86,22 +120,43 @@ public synchronized List<BlockHeader> getFirstNAncestorHeaders(final int size) {
   }
 
   public synchronized void prependAncestorsHeader(final BlockHeader blockHeader) {
-    if (firstStoredAncestor.isEmpty()) {
-      firstStoredAncestor = Optional.of(blockHeader);
-      lastStoredPivot = Optional.of(blockHeader);
+    prependAncestorsHeader(blockHeader, false);
+  }
+
+  public synchronized void prependAncestorsHeader(
+      final BlockHeader blockHeader, final boolean alreadyStored) {
+    if (!alreadyStored) {
```

Comment:
Don't think this condition is covered by tests - if you always call headers.put everything still passes.

Reply:
this is only a for performance, in the case we already know that the header is already present, because we just read it, so we can avoid writing to the db